(Homework №8): Template and showcards prompt

Abdulhamidovich opened the pull request.

Abdulhamidovich committed:
homework-8: template and showcards prompt

pocahontassss commented on product-cards.js:
```diff
@@ -0,0 +1,77 @@
+        compound: [
+            'активные натуральные комплексы',
+            'витамины С, А, РР, В и Е',
+            'солнцезащитные компоненты'],
```
скобка на новую строку, таковы правила форматирования кода

pocahontassss commented on product-cards.js:
```diff
@@ -0,0 +1,77 @@
+    },
+    {
+        productCategory: 'для нормальной кожи',
+        productName: 'Увлажняющая маска',
+        description: 'Способствует удерживанию влаги в верхних слоях кожи.',
+        compoundLabel: 'Состав',
+        compound: [
+            'Воски',
+            'Минералы',
+            'Масла'],
+        priceLabel: 'Цена:',
+        price: 3500,
+        imgSrc: '/images/product-mask.png',
+        imgAlt: 'Увлажняющая маска',
+        currency: '₽'
+    },
```
https://t.me/c/2614739958/5341/10503

20:15 - 37:30 — Как правильно составлять объекты на основе данных.  Разбираем на примере ДЗ и Алиэкспресса

Посмотри этот отрезок. Есть проблема с объектом. Есть лишние свойства, не относящиеся к продукту (priceLabel, compoundLabel). также imgAlt дублирует название продукта. И убери вообще слово "product" из объекта, мы понимаем, что это продукт исходя из названия массива

pocahontassss commented on product-cards.js:
```diff
@@ -0,0 +1,77 @@
+            'Масла'],
+        priceLabel: 'Цена:',
+        price: 3500,
+        imgSrc: '/images/product-mask.png',
```
здесь будем вставлять лишь название картинки

pocahontassss commented on homework-8.js:
```diff
@@ -0,0 +1,49 @@
+import {productCardTemplate} from "./product- cards.js";
```
ошибка синтаксическая, у тебя не сработает код, убери пробел

pocahontassss commented on homework-8.js:
```diff
@@ -0,0 +1,49 @@
+        return cardsCount();
+    }
+    return result;
+}
```
**Правила именования функций/методов:**

1. Название начинается с маленькой буквы, все последующие слова - с большой. (стиль camelCase).
2. Название начинается с глагола. Разрешено добавить существительное для уточнения того, над чем именно выполняется действие.
Функция — это действие. Имя должно отвечать на вопрос “что сделать”, а не “что вернуть”.
 
Примеры: 
- getUserInfo (получить информацию о пользователе);
- showMessage   (отобразить сообщение);
- formatDate   (форматировать дату);
- calculateValue   (рассчитать значение);
- saveSettings   (сохранить настройки);
- deleteProfile  (удалить профиль):
- login   (авторизоваться);
- update, create, remove, load и т.д;

3. В большинстве случаев, если функция что-то возвращает (return) - используется префикс get.

Примеры:
- getUserName (получить имя пользователя);
- getSettings (получить настройки);
- getProduct (получить продукт);

4. Если функция возвращает логический тип (boolean) - используются префиксы: is, has, can, should.

Примеры:
- canEdit (можно редактировать);
- isAdmin (это админ):
- hasAccess (имеется доступ);
- shouldOpen (должен быть открыт);

5. Если функция обрабатывает событие - используются префиксы: handle, on.

Примеры:
- onClick (при клике);
- onSubmit (при отправке);
- onHover (при наведении);
- handleDelete (обработка удаления);
- handleClose (обработка закрытия);

pocahontassss commented on homework-8.js:
```diff
@@ -0,0 +1,49 @@
+    cardCloned.querySelector('.product-price-label').textContent = card.priceLabel;
+    cardCloned.querySelector('.product-price').textContent = `${card.price} ${card.currency}`;
+    cardCloned.querySelector('img').src = card.imgSrc;
+    cardCloned.querySelector('img').alt = card.imgAlt;
```
это мы можем взять из названия карточки

pocahontassss commented on homework-8.js:
```diff
@@ -0,0 +1,49 @@
+    cardCloned.querySelector('.product-category').textContent = card.productCategory;
+    cardCloned.querySelector('.product-name').textContent = card.productName;
+    cardCloned.querySelector('.description').textContent = card.description;
+    cardCloned.querySelector('.compound-label').textContent = card.compoundLabel;
```
это можно вынести в html (template)

pocahontassss commented on homework-8.js:
```diff
@@ -0,0 +1,49 @@
+    cardCloned.querySelector('.description').textContent = card.description;
+    cardCloned.querySelector('.compound-label').textContent = card.compoundLabel;
+    cardCloned.querySelector('.product-compound').textContent = card.compound;
+    cardCloned.querySelector('.product-price-label').textContent = card.priceLabel;
```
это можно вынести в html (template)

pocahontassss commented on homework-8.js:
```diff
@@ -0,0 +1,49 @@
+    cardCloned.querySelector('.product-name').textContent = card.productName;
+    cardCloned.querySelector('.description').textContent = card.description;
+    cardCloned.querySelector('.compound-label').textContent = card.compoundLabel;
+    cardCloned.querySelector('.product-compound').textContent = card.compound;
```
здесь ты должен создавать li'шки, а не обычный текстовый контент. Посмотри лекцию выше, там об этом говорю

Commit:
fixed errors

# Conflicts:
#	homework-8.js
#	index.html
#	product-cards.js
#	style.css

anatolyeviich commented on product-cards.js:
```diff
@@ -74,4 +75,72 @@ export const productCardTemplate = [
+            'солнцезащитные компоненты'
+        ],
+        price: 2500,
+        imgSrc: 'product-muss.png',
```
src - это путь к картинке, а мы здесь указываем только ее название

.png - это не название, а формат картинки. Исправь

anatolyeviich commented on homework-8.js:
```diff
@@ -1,49 +1,53 @@
+    cardCloned.querySelector('img').src = `${imgSrc}${card.imgSrc}`;
+    cardCloned.querySelector('img').alt = card.name;
+    cardList.appendChild(cardCloned);
+});
```
форматирование исправь

anatolyeviich commented on homework-8.js:
```diff
@@ -1,49 +1,53 @@
+
+const showCards = getCardsCount()
+productCardTemplate.forEach((card, index) => {
+    if (index >= showCards) return;
```
код работает некорректно. Ты спрашиваешь у пользователя, сколько карточек ему отобразить и на основе этой информации отображаешь нужное кол-во

anatolyeviich commented on homework-8.js:
```diff
@@ -1,49 +1,53 @@
+productCardTemplate.forEach((card, index) => {
+    if (index >= showCards) return;
+    const cardCloned = cardListTemplate.content.cloneNode(true);
+    const imgSrc = '/images/'
```
нет смысла его сохранять в переменную, ты вызываешь ее всего-лишь 1 раз, можно ручками написать

anatolyeviich commented on homework-8.js:
```diff
@@ -1,49 +1,53 @@
+    if (index >= showCards) return;
+    const cardCloned = cardListTemplate.content.cloneNode(true);
+    const imgSrc = '/images/'
+    const compaundList = item => {
```
compound*

anatolyeviich reviewed (changes requested):
Я не уверен, что твой код вообще работает как полагается

Abdulhamidovich committed:
fix: resolve code issues

- correct image paths in product-cards.js
- fix card display logic in homework-8.js
- rename variables to proper naming
- improve code formatting

pocahontassss commented on images/bg_flower.png:
Нарушены правила наименования изображений. 
Подробнее о рекомендуемых практиках можно прочитать в статье – [[ссылка]](https://docs.google.com/document/u/0/d/1ipRxAtW0KEC5bj-vWLNk-p8AwjsN0AWmN8YApyP9VJw/mobilebasic)

pocahontassss commented on homework-8.js:
```diff
@@ -0,0 +1,50 @@
+    cardCloned.querySelector('img').src =`/images/${card.imgName}.png`;
+    cardCloned.querySelector('img').alt = card.name;
+    cardList.appendChild(cardCloned);
+});
```
и снова таки, данный код работает некорректно. Ты проверял его работоспособность? Он запускается вообще? 
Создай функцию, которая генерирует карточки на основе переданного массива, перечитай задание

Abdulhamidovich committed:
fix image names and add function

pocahontassss commented on homework-8.js:
```diff
@@ -0,0 +1,54 @@
+    });
+}
+
+renderProductCards(productCardTemplate)
```
ты должен в функцию передавать массив карточек, которые хочешь отобразить, и всё. Функция принимает массив и обрабатывает его, а вычислять массив ты будешь уже в getCardsCount. Перепиши

Abdulhamidovich committed:
fix: function fixed

pocahontassss commented on homework-8.js:
```diff
@@ -0,0 +1,54 @@
+};
+
+const renderProductCards = (template) => {
+    const maxCards = getCardsCount();
+    template.forEach((card, index) => {
+        if (index >= maxCards) return;
+        const cardCloned = cardListTemplate.content.cloneNode(true);
+        cardCloned.querySelector('.product-category').textContent = card.category;
+        cardCloned.querySelector('.product-name').textContent = card.name;
+        cardCloned.querySelector('.description').textContent = card.description;
+        cardCloned.querySelector('.product-compound').innerHTML =
+            card.compound
+                .map(item => `<li>${item}</li>`)
+                .join('');
+        cardCloned.querySelector('.product-price').textContent = `${card.price} ${card.currency}`;
+        cardCloned.querySelector('img').src =`/images/${card.imgName}.png`;
+        cardCloned.querySelector('img').alt = card.name;
+        cardList.appendChild(cardCloned);
+    });
+}
```
еще раз. Напиши функцию, которая принимает массив карточек (это может быть массив из 5, из 2, или другого кол-ва карточек) и просто рендерит их (отрисовывает), больше ничего

pocahontassss replied:
А значит высчитывать кол-во ты будешь вне этой функции

Abdulhamidovich committed:
fix: card rendering function corrected
